# Sowing Success: How Machine Learning Helps Farmers Select the Best Crops

![Farmer in a field](farmer_in_a_field.jpg)

Measuring essential soil metrics such as nitrogen, phosphorous, potassium levels, and pH value is an important aspect of assessing soil condition. However, it can be an expensive and time-consuming process, which can cause farmers to prioritize which metrics to measure based on their budget constraints.

Farmers have various options when it comes to deciding which crop to plant each season. Their primary objective is to maximize the yield of their crops, taking into account different factors. One crucial factor that affects crop growth is the condition of the soil in the field, which can be assessed by measuring basic elements such as nitrogen and potassium levels. Each crop has an ideal soil condition that ensures optimal growth and maximum yield.

A farmer reached out to you as a machine learning expert for assistance in selecting the best crop for his field. They've provided you with a dataset called `soil_measures.csv`, which contains:

- `"N"`: Nitrogen content ratio in the soil
- `"P"`: Phosphorous content ratio in the soil
- `"K"`: Potassium content ratio in the soil
- `"pH"` value of the soil
- `"crop"`: categorical values that contain various crops (target variable).

Each row in this dataset represents various measures of the soil in a particular field. Based on these measurements, the crop specified in the `"crop"` column is the optimal choice for that field.  

In this project, you will build multi-class classification models to predict the type of `"crop"` and identify the single most importance feature for predictive performance.

In [8]:
# All required libraries are imported here for you.
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score
from sklearn.preprocessing import StandardScaler


# Load the dataset
crops = pd.read_csv("soil_measures.csv")

# Write your code here
X = crops.drop("crop", axis = 1)
y = crops["crop"]

best_feature = None
best_score = 0
best_precision = 0 
best_recall = 0 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state= 42)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Iterate through each feature
for feature in X.columns:  
    col_idx = X.columns.get_loc(feature)  # Obtiene la posición de la columna en el DataFrame original
    
    X_train_single = X_train_scaled[:, col_idx].reshape(-1, 1)  
    X_test_single = X_test_scaled[:, col_idx].reshape(-1, 1) 
    
    
    model = LogisticRegression(max_iter=200)
    model.fit(X_train_single, y_train)

  # Predict and evaluate performance
    y_pred = model.predict(X_test_single)
    score = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average = 'weighted' , zero_division = 0)
    recall = recall_score(y_test, y_pred, average = 'weighted', zero_division = 0)
    
    # Update best feature if current one performs better
    if score > best_score:
        best_feature = feature
        best_score = score
        best_precision = precision 
        best_recall = recall 

# Store result in dictionary
best_predictive_feature = {
    "features": best_feature, 
    "accuracy": best_score, 
    "precision" : best_precision, 
    "recall": best_recall 
}

# Output the best feature and its score
print(best_predictive_feature)



{'features': 'K', 'accuracy': 0.24545454545454545, 'precision': 0.11116577536331361, 'recall': 0.24545454545454545}


# CONCLUSIÓN 

El análisis realizado con regresión logística muestra que la mejor característica predictiva para determinar el tipo de cultivo es el nivel de potasio ("K") en el suelo. Sin embargo, los resultados indican que el modelo tiene un rendimiento bajo:

Exactitud (Accuracy): 24.5% → El modelo solo acierta en el 24.5% de los casos, lo que sugiere que no está capturando bien los patrones en los datos.

Precisión (Precision): 11.11% → Muchas de las predicciones son incorrectas, indicando un alto número de falsos positivos.

Exhaustividad (Recall): 24.5% → El modelo no está identificando correctamente todas las instancias de cada cultivo.